src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs

```diff
@@ -127,14 +130,19 @@ private static unsafe OperationStatus DecodeFromUtf8(ReadOnlySpan<byte> utf8, Sp
                     // Therefore, (destLength / 3) * 4 will always be less than 2147483641
                     Debug.Assert(destLength < (int.MaxValue / 4 * 3));
                     maxSrcLength = (destLength / 3) * 4;
+                    int remainder = destLength % 3;
+                    if (isFinalBlock && remainder > 0)
+                    {
+                        srcLength &= ~0x3; // In case of Base64UrlDecoder source can be not a multiple of 4, round down to multiple of 4
+                    }
```

Comment:
Note that this branch only hit when destination is not enough, when destination is enough  (`destLength >= decodedLength`) `maxSrcLength = srcLength - skipLastChunk`

As you can see `maxSrcLength = (destLength / 3) * 4` here is not accounting the fraction part `destLenght % 3`. I tried to account the faction part instead of removing fraction part from source, but that causes the below `while (src < srcMax)` condition to true, which means the loop gonna work even there were less than 4 bytes available in the source. So instead, I am rounding down the `srcLength`. So basically this is done for:
1.  `while (src < srcMax)`
2. `if (maxSrcLength != srcLength - skipLastChunk) goto DestinationTooSmallExit;`

I am not sure what the Result column express in your table, but looks the table needs small corrections for some cell:
srcLength | destLength | isFinal | maxDecode | skipLength | enough | maxSrcLength | src-skip | DTSO | remainder | src&=~3 | src3-skip | DTS1
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
16 | 12 | TRUE | 12 | 4 | TRUE | 12 | 12 | FALSE | 0 | 16 | 12 | TRUE
16 | 12 | FALSE | 12 | 0 | TRUE | 16 | 16 | FALSE | 0 | 16 | 16 | FALSE
17 | 12 | TRUE | 12 | 4 | TRUE | 16 | 13 | TRUE | 0 | 16 | 12 | TRUE
17 | 12 | FALSE | 12 | 0 | TRUE | 16 | 17 | TRUE | 0 | 16 | 16 | FALSE
18 | 12 | TRUE | 13 | 4 | FALSE | 16 | 14 | TRUE | 0 | 16 | 12 | TRUE
18 | 12 | FALSE | 13 | 0 | FALSE | 16 | 18 | TRUE | 0 | 16 | 16 | FALSE
19 | 12 | TRUE | 14 | 4 | FALSE | 16 | 15 | TRUE | 0 | 16 | 12 | TRUE
19 | 12 | FALSE | 14 | 0 | FALSE | 16 | 19 | TRUE | 0 | 16 | 16 | FALSE

Reply:
> I am not sure what the Result column express in your table

Result: `=IF(OR(F2, NOT(C2), J2 = 0), I2, M2)` (Choose DTS0 or DTS1 based on the two ifs that control whether we do "&~3")

> instead of removing the fraction part from srcLength, but that causes the below `while (src < srcMax)` condition to true,

Changing `srcLength` here doesn't change `src` or `srcMax (= srcBytes + maxSrcLength)`, so I don't see how behavior of the loop would be affected by this.

Reply:
> Changing `srcLength` here doesn't change src or srcMax (= srcBytes + maxSrcLength), so I don't see how behavior of the loop would be affected by this.

Yes, `srcLength` doesn't affect `src` or `srcMax`, I meant, accounting/adding `destLenght % 3` fraction to `maxSrcLength` would cause `src < srcMax` because `srcMax = srcBytes + maxSrcLength`, the while loop requires whole 4 bytes, therefore keeping the `maxSrcLength` multiple of 4, and round down the `srcLength` into multiple of 4 int order to make `if (maxSrcLength != srcLength - skipLastChunk) goto DestinationTooSmallExit;` work correctly

Reply:
I fixed my error in computing maxDecode (it was right on the whiteboard before I made the spreadsheet :smile:), and added a column which explains whether we're picking DTS0 (unaltered) or DTS1 (using the round-down thing).  I only found one input that even picks DTS1, and the DTS0 and DTS1 calculations are the same:

![image](https://github.com/dotnet/runtime/assets/10642668/6fafa1df-6287-43d1-be51-5cbcba337313)

Reply:
Looks `maxSrcLength` in the table calculated incorrectly from row 4, when `destLength` is 12, 13 or 14 it will always equal to 16 (`maxSrcLength = (destLength / 3) * 4`)

Reply:
> I'm not finding a set of inputs that will change the behavior on the first if. Can you provide one to me?

Here is my version of the list:
<img width="499" alt="image" src="https://github.com/dotnet/runtime/assets/45579687/55bee3fb-99d1-4dce-814a-340106fcd9fc">

Some columns values are different than yours, even though the final results same as yours for the numbers you tested, below in green are the values that makes difference in the behavior

Reply:
Some of your `src&=~3` are wrong.  e.g. in (17, 12, true) you get 17, but it's 16.

I do agree, though, that (11, 5, true) is a case where we change the truth state of `if (maxSrcLength != srcLength - skipLastChunk)` on what is currently line 157.  So, I worked it out.

<details>
  <summary>Working it out by hand...</summary>

```
ReadOnlySpan<T> utf8 = stackalloc T[11];
Span<byte> bytes = stackalloc byte[5];
bool isFinalBlock = true;

if (utf8.IsEmpty) => false

T* srcBytes = utf8;
T* destBytes = bytes;
int srcLength = 11;
int destLength = 5;
int maxSrcLength = 11;
int decodedLength = (srcLength / 4 * 3 + Math.Max(0, (srcLength % 4) - 1) => (11 / 4 * 3 + Max(0, (11 % 4) - 1)) => (2 * 3 + Max(0, 3 - 1)) => (6 + Max(0, 2)) => 8

if (destLength < decodedLength - 2) => true

maxSrcLength = destLength / 3 * 4 => 5 / 3 * 4 => 1 * 4 => 4

T* src = srcBytes;
byte* dest = destBytes;
T* srcEnd = srcBytes + 11
T* srcMax = srcBytes + 4

if (maxSrcLength >= 24) => false

int skipLastChink = isFinalBlock ? 4 : 0 => 4

if (destLength >= decodedLength) => false
(maxSrcLength, int remainder) = DivRem(destLength, 3) => DivRem(5, 3) => (1, 2)
maxSrcLength *= 4 => maxSrcLength = 4

if (isFinalBlock && remainder > 0) => true

srcLength &= 0x3 => srcLength = 11 & ~3 => srcLength = 8

ref sbyte decodingMap = ...
srcMax = srcBytes + maxSrcLength => srcBytes + 4

143:while (src < srcMax) => true

int result = TDecoder.Decode(src, ref decodingMap)

if (result < 0) => false

WriteThreeLowOrderBytes(dest, result)
src += 4 => src = srcBytes + 4
dest += 3 => dest = destBytes + 3

143:while (src < srcMax) => false

if (maxSrcLength != srcLength - skipLastChunk) => if (4 != 8 - 4) => if (4 != 4) => false

if (src == srcEnd) => if (srcBytes + 4 == srcBytes + 11) => if (4 == 11) => false

long remaining = srcEnd - src => 11 - 4 => 7
int i0 = TDecoder.DecodeRemaining(srcEnd, out uint t2, out uint t3)
byte* destMax = destBytes + destLength => destBytes + 5

if (!TDecoder.IsValidPadding(t3)) => ??? assume true ... return DestinationTooSmall.  So assume false.

if (!TDecoder.IsValidPadding(t2)) => ??? assume true ...  dest += 2; src += remaining => src = srcBytes + 4 + 7 => srcBytes + 11
   => assume false ... dest += 1; src += remaining => src = srcBytes + 4 + 7 => srcBytes + 11

if (srcLength != utf8.Length) => if (srcBytes + 11 != srcBytes + 11) => if (11 != 11) => false

bytesConsumed = (src - srcBytes) => (srcBytes + 11 - srcBytes) => 11
bytesWritten = (dest - destBytes) => (destBytes + 3 + 2 - destBytes) => 5
return OperationStatus.Done
```
</details>

* `DecodeFrom("AAAAAAAAAAA", new byte[5], true)` will produce DestinationTooSmall on line 202, because no padding was found in the second chunk of 4 and dest can't hold 3 more bytes.
* `DecodeFrom("AAAAAAA====", new byte[5], true)`, if nothing already rejected it for having too many padding bytes, will result in Done.  But it better be rejected as InvalidData somewhere for too many padding bytes
* I can't come up with an input on (11, 5) that isn't DestinationTooSmall, because it clearly needs to write at least two complete blocks.  Unless whitespace hasn't been taken out here, in which case I have questions.

Can you 1) please demonstrate for me what test would pass if you deleted this line of code, and 2) add it as a comment.

All I want is something like

```suggestion
                    if (isFinalBlock && remainder > 0)
                    {
                        // Trim this down so that "AAAAAAAAAAA" (11 chars) into `new byte[5]`
                        // reports _______ instead of DestinationTooSmall for Base64Url.
                        srcLength &= ~0x3;
                    }
```

except with the actual values swapped in.

Reply:
> I can't come up with an input on (11, 5) that isn't DestinationTooSmall, because it clearly needs to write at least two complete blocks. Unless whitespace hasn't been taken out here, in which case I have questions.

Your evaluations all correct, as you noted white space is the 3rd case, on InvalidData exit it will call `InvalidDataFallback` that would detect and slice the whitespace and recursively call back the `DecodeFrom<,>(...)` again with valid length and valid data.
https://github.com/dotnet/runtime/blob/b871d6db15d65a3180094f7b5eeeeefbab0ccfb3/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs#L272-L277

> Can you 1) please demonstrate for me what test would fail if you deleted this line of code, and 2) add it as a comment.

```suggestion
                    if (isFinalBlock && remainder > 0)
                    {
                        // Trim this down so that "AAAA\t\t\tAAA=" (11 chars) into `new byte[5]`
                        // reports Done instead of DestinationTooSmall for Base64Url.
                        srcLength &= ~0x3; multiple of 4
                    }
```

JFYI the exact unit test that covers these scenarios:

https://github.com/dotnet/runtime/blob/b871d6db15d65a3180094f7b5eeeeefbab0ccfb3/src/libraries/System.Memory/tests/Base64Url/Base64UrlDecoderUnitTests.cs#L721-L726